client/asset/btc/txdb.go

```diff
@@ -0,0 +1,408 @@
+	binary.BigEndian.PutUint64(key[len(pendingPrefix):], i)
+	return key
+}
+
+// blockKey maps a block height and an index to an extendedWalletTransaction.
+// The index is required as there may be multiple transactions in the same
+// block.
+func blockKey(blockHeight, index uint64) []byte {
+	key := make([]byte, len(blockPrefix)+16)
+	copy(key, blockPrefix)
+	binary.BigEndian.PutUint64(key[len(blockPrefix):], blockHeight)
+	binary.BigEndian.PutUint64(key[len(blockPrefix)+8:], index)
+	return key
+}
+
+func parseBlockKey(key []byte) (blockHeight, index uint64) {
+	blockHeight = binary.BigEndian.Uint64(key[len(blockPrefix):])
+	index = binary.BigEndian.Uint64(key[len(blockPrefix)+8:])
+	return
+}
+
+// txKey maps a txid to a blockKey or pendingKey.
+func txKey(txid []byte) []byte {
+	key := make([]byte, len(txPrefix)+len(txid))
+	copy(key, txPrefix)
+	copy(key[len(txPrefix):], txid)
+	return key
+}
+
+type txDB interface {
+	storeTx(tx *extendedWalletTx) error
+	markTxAsSubmitted(txID dex.Bytes) error
+	getTxs(n int, refID *dex.Bytes, past bool) ([]*asset.WalletTransaction, error)
+	getPendingTxs() ([]*extendedWalletTx, error)
+	removeTx(hash dex.Bytes) error
+	// setLastReceiveTxQuery stores the last time the wallet was queried for
+	// receive transactions. This is required to know how far back to query
+	// for incoming transactions that were received while the wallet is
+	// offline.
+	setLastReceiveTxQuery(block uint64) error
+	getLastReceiveTxQuery() (uint64, error)
+	close() error
+	run(context.Context)
+}
+
+type badgerTxDB struct {
+	*badger.DB
+	log dex.Logger
+}
+
+var _ txDB = (*badgerTxDB)(nil)
+
+// badgerLoggerWrapper wraps dex.Logger and translates Warnf to Warningf to
+// satisfy badger.Logger.
+type badgerLoggerWrapper struct {
+	dex.Logger
+}
+
+var _ badger.Logger = (*badgerLoggerWrapper)(nil)
+
+// Warningf -> dex.Logger.Warnf
+func (log *badgerLoggerWrapper) Warningf(s string, a ...interface{}) {
+	log.Warnf(s, a...)
+}
+
+func newBadgerTxDB(filePath string, log dex.Logger) (*badgerTxDB, error) {
+	// If memory use is a concern, could try
+	//   .WithValueLogLoadingMode(options.FileIO) // default options.MemoryMap
+	//   .WithMaxTableSize(sz int64); // bytes, default 6MB
+	//   .WithValueLogFileSize(sz int64), bytes, default 1 GB, must be 1MB <= sz <= 1GB
+	opts := badger.DefaultOptions(filePath).WithLogger(&badgerLoggerWrapper{log})
+	db, err := badger.Open(opts)
+	if err == badger.ErrTruncateNeeded {
+		// Probably a Windows thing.
+		// https://github.com/dgraph-io/badger/issues/744
+		log.Warnf("newTxHistoryStore badger db: %v", err)
+		// Try again with value log truncation enabled.
+		opts.Truncate = true
+		log.Warnf("Attempting to reopen badger DB with the Truncate option set...")
+		db, err = badger.Open(opts)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return &badgerTxDB{
+		DB:  db,
+		log: log}, nil
+}
+
+func (db *badgerTxDB) findFreeBlockKey(txn *badger.Txn, blockNumber uint64) ([]byte, error) {
```

Comment:
I could have sworn I already commented on this, but apparently not.

Why not just use the tx's block index?

https://github.com/buck54321/dcrdex/compare/0cc91cc...buck54321:dcrdex:indexsequence

Reply:
I made the change, then I realized that btcwallet is not returning the block index. I guess we could update btcwallet, ltcwallet, and bchwallet, but is it really worth it? We have to do the this free key finding loop for pending transactions anyways.

Reply:
Oof. No. Just use `findFreeBlockKey`.